packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,378 @@
+        setAttribute: {
+            value: function setAttribute(name: string, value: any) {
+                if (offset.has(name)) {
+                    const old = nativeGetAttribute.call(this, name);
+                    // maybe we want to call the super.setAttribute rather than the native one
+                    nativeSetAttribute.call(this, name, value);
+                    attributeChangedCallback.call(this, name, old, value + '');
+                } else {
+                    nativeSetAttribute.call(this, name, value);
+                }
+            },
+            writable: true,
+            enumerable: true,
+            configurable: true,
+        },
+        removeAttribute: {
+            value: function removeAttribute(name: string) {
+                if (offset.has(name)) {
+                    const old = nativeGetAttribute.call(this, name);
+                    // maybe we want to call the super.removeAttribute rather than the native one
+                    nativeRemoveAttribute.call(this, name);
+                    attributeChangedCallback.call(this, name, old, null);
+                } else {
+                    nativeRemoveAttribute.call(this, name);
+                }
+            },
+            writable: true,
+            enumerable: true,
+            configurable: true,
+        },
+    });
+}
+
+// Helper to upgrade an instance with a CE definition using "constructor call trick"
+function internalUpgrade(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instancedDefinition: Definition
+) {
+    Object.setPrototypeOf(instance, instancedDefinition.UserCtor.prototype);
+    definitionForElement.set(instance, instancedDefinition);
+    // attributes patches when needed
+    if (instancedDefinition !== registeredDefinition) {
+        patchAttributes(instance, registeredDefinition, instancedDefinition);
+    }
+    // Tricking the construction path to believe that a new instance is being created,
+    // that way it will execute the super initialization mechanism but the HTMLElement
+    // constructor will reuse the instance by returning the upgradingInstance.
+    // This is by far the most important piece of the puzzle
+    upgradingInstance = instance;
+    new instancedDefinition.UserCtor();
+
+    const { observedAttributes, attributeChangedCallback } = instancedDefinition;
+    if (observedAttributes.size === 0 || attributeChangedCallback === undefined) {
+        return;
+    }
+    const offset = getObservedAttributesOffset(registeredDefinition, instancedDefinition);
+    if (offset.size === 0) {
+        return;
+    }
+    // Approximate observedAttributes from the user class, but only for the offset attributes
+    offset.forEach((name) => {
+        if (nativeHasAttribute.call(instance, name)) {
+            const newValue = nativeGetAttribute.call(instance, name);
+            attributeChangedCallback.call(instance, name, null, newValue);
+        }
+    });
```

Comment:
Would it make sense for this `internalUpgrade` function to handle properties set on the HTML element before it's upgraded? This is something [I've run into with web components before](https://nolanlawson.com/2021/08/03/handling-properties-in-custom-element-upgrades/).